### PR TITLE
Added M118 message forwarding

### LINKED ISSUE #2572: [FR] Send Pause and Stop Command While Remote Printing
Would be really nice to be able to use the pause and stop key when remote printing.  This could be set up to send a M601 that would pause Octoprint with the right config.

Thanks,
Ryan

### LINKED ISSUE #2618: Execute command for Nextion with M118 gcode.
Hi,
I am trying to control Nextion device with Skr Pro v1.2 But I couldn't find a way. 

I enabled  #define NEXTION_TFT nextion module and connected to TFT port. I used many versions of M118 codes but it didn't work.

Fo ex: M118 P2 page page2 or M118 A1 P2 page page2 or M118 P2 page page2 M118 P2 0xFF 0xFF 0xFF or whatever else....

These are serial settings. 
#define SERIAL_PORT -1
#define BAUDRATE 250000

#define SERIAL_PORT_2 1
#define BAUDRATE_2 9600   // Enable to override BAUDRATE

With this way, i can control SKr Pro with the Nextion device. For ex : print "M104 S210 T0"  it is working properly.

Now I need a hand to control pages or values on Nextion device with M118 command or another way.

Thank you from now

### LINKED ISSUE #2638: [BUG] M118 Commands do not go through to ESP3D when using repo code (not release). M118 commands work fine in all other scenarios
### Description

M118 Commands do not go through to ESP3D when using repo code (not release).  M118 commands work fine in all other scenarios

Example:
M118 P0 [ESP201] P23 V254 ANALOG=YES 
**Should be received by ESP3d and set pin 23 to full power**

Source:
ESP3D WebUI:  Passed
Onboard SD:  Passed
Motherboard custom commands: Passed
TFT SD:  Fail
TFT USB:  Fail
TFT Custom Commands:  Fail
TFT Terminal:  Cannot test due to no brackets in terminal

### Steps to reproduce

1. Compile and load master branch on TFT
2. Place gcode for M118 command in any of the failed locations mentioned above
3. Run gcode

**Expected behavior**
Command is sent to all serial ports and executes on ESP3D

**Actual behavior**
Command never reaches ESP3D.  No output

### Hardware Variant
BTT TFT35 E3 V3.0
### TFT Firmware Version & Main Board Firmware details
Master branch, Marlin 2.1.1, ESP3D 2.1 & 3.0

### Additional Information

[Configuration.zip](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/10065594/Configuration.zip)

### LINKED ISSUE #2640: [BUG] Screen freeze on onboard SD print abort
### Description

The screen will be stuck at 'Loading...' or 'busy processing, please wait' when cancelling an onboard SD print. The message that shows depends on the gcode that is run. See below. This does not occur when printing from the TFT card slot.

### Steps to reproduce

1. Upload the following Gcode to an SD card:
G28
M109 S100
2. Insert the SD card into the motherboard.
3. Start the print.
4. Cancel the print during heating.
5. Press 'Back' button
6. Blank screen stuck showing 'Loading...' 

### Another way to reproduce

1. Upload the following Gcode:
G28
M140 S60
M105
M190 S60
M104 S100
M105
M109 S100
2. Insert the SD card into the motherboard.
3. Start the print.
3. Cancel the print during bed heating.
4. Screen freezes showing 'busy processing, please wait' 

**Expected behavior**
Run cancel gcode.

**Actual behavior**
Screen becomes frozen, does not run cancel gcode.

### Hardware Variant
B1 TFT35 V3.0.1
SKR MINI E3 V3.0
### TFT Firmware Version & Main Board Firmware details
Marlin Firmware: 2.1.1
TFT Firmware: BIGTREETECH_GD_TFT35_B1_v3.0 V27.x Nov 10 2022

### Additional Information
[bug.zip](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/10079953/bug.zip)

### PR DESCRIPTION
**Pull request by digant73**
**IMPROVEMENTS:**
* Added M118 message forwarding: Any M118 command read by TFT (e.g. printing from TFT or from remote host (e.g. OctoPrint etc.)) is parsed and forwarded to the provided TFT destination port/s, if enabled.
  This feature allowed an easy implementation (just an M118 message) for:
  * FR #2572 (reported below as next improvement)
  * FR #2618
  * FR #2638

  The new `Serial_Forward` function was added to the existing `SerialConnection` API
* Implemented #2572: Now, as also reported in README.md, when on Printing menu, pressing on the **pause**, **resume** and **stop** buttons will trigger to the remote host the following notifications, respectively:

  | **NOTIFICATION**            | **ACK MESSAGE**                       |
  | :-------------------------- | :------------------------------------ |
  | **pause**                   | `//action:notification remote pause`  |
  | **resume**                  | `//action:notification remote resume` |
  | **cancel**                  | `//action:notification remote cancel` |

  The remote host must properly handle the received notifications. For example, if `//action:notification remote pause` is received then the remote host must effectively pause the print and send `M118 P0 A1 action:pause` in order to trigger the pause action to the TFT.

* Serial functions `syncL2CacheFromL1` and `parseRcvGcode` moved from `parseAck.c` API to `SerialConnection.c` API and renamed as the following, respectively:
  * `Serial_Get` (retrieve a message)
  * `Serial_GetFromUART` (retrieve messages from all the enabled supplementary ports storing them in the command queue (in interfaceCmd.c) for further processing)
* Serial functions in HAL's `Serial` API properly renamed (common naming convention) as:
  * `Serial_Put` (send a message)
  * `Serial_PutChar` (send only a char)
* Beeing part of the main changes of the PR, some code reduction and cleanup has been also provided in `interfaceCmd.c`:
  * `infoCmd` data queue renamed to `cmdQueue` (by definition the word `cmd` (command) means a set of info). Used `cmdQueue` instead on `cmd_queue` due it is not a primitive type (e.g. an array of chars) but a data structure (the use of `_` and further `.` (for an attribute) would result in a less readable text
  * Grouped M25, M125 and M524 (code reduction)
  * Grouped M105 and M155 (code reduction)
  * Some code sharing, reduction and fixes applied on top of last merged PR #2607. Also, no need  to handle M108, M112 etc. as emergency commands when sent from command queue (Marlin will always replies with an "ok")
* Beeing part of the main changes of the PR, some code reduction and cleanup has been also provided in `parseAck.c` (e.g. `dmaL2Cache` buffer renamed to a more understandable `ack_cache` (an array of chars) etc.).

fixes #2640
resolve #2572
resolve #2618
resolve #2638

**PR STATE:** Ready for merge

### CODE REVIEW

File: TFT/src/User/API/interfaceCmd.c [hunk 1]
```diff
@@ -25,37 +25,35 @@ typedef enum
...
+  for (int i = 0; i < cmdQueue.count; i++)
   {
-    found = strcmp(cmd, infoCmd.queue[(infoCmd.index_r + i) % CMD_QUEUE_SIZE].gcode) == 0;
+    if (strcmp(cmd, cmdQueue.queue[(cmdQueue.index_r + i) % CMD_QUEUE_SIZE].gcode) == 0)
+      return true;
   }
 
-  return found;
+  return false;
 }
```
Comment (kisslorand) [inline TFT/src/User/API/interfaceCmd.c:58]: Stolen from PR #2607 
```
bool isEnqueued(const CMD cmd)
{
  uint8_t i = 0;

  while (i < infoCmd.count)
  {
    if (strcmp(cmd, infoCmd.queue[(infoCmd.index_r + i++) % CMD_QUEUE_SIZE].gcode) == 0)
      return true;
  }

  return false;
}
```

File: TFT/src/User/API/interfaceCmd.c [hunk 2]
```diff
@@ -466,11 +462,10 @@ void syncTargetTemp(uint8_t index)
   }
 }
 
-// Parse and send gcode cmd in infoCmd queue.
+// Parse and send gcode cmd in cmdQueue queue.
 void sendQueueCmd(void)
 {
-  if (infoHost.wait == true) return;
-  if (infoCmd.count == 0) return;
+  if (infoHost.wait == true || cmdQueue.count == 0) return;
```
Comment (kisslorand) [inline TFT/src/User/API/interfaceCmd.c:468]: Copy & Paste from PR #2607

File: TFT/src/User/API/interfaceCmd.c [hunk 3]
```diff
@@ -523,8 +518,8 @@ void sendQueueCmd(void)
     case 'M':
       switch (cmd_value())
       {
-        case 0:
-        case 1:
+        case 0:  // M0
```
Comment (kisslorand) [inline TFT/src/User/API/interfaceCmd.c:521]: Copy & Paste from PR https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/pull/2607

File: TFT/src/User/API/interfaceCmd.c [hunk 4]
```diff
@@ -523,8 +518,8 @@ void sendQueueCmd(void)
...
+        case 1:  // M1
```
Comment (kisslorand) [inline TFT/src/User/API/interfaceCmd.c:522]: Copy & Paste from PR https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/pull/2607

File: TFT/src/User/API/interfaceCmd.c [hunk 5]
```diff
@@ -735,63 +732,57 @@ void sendQueueCmd(void)
...
+          }
 
         #else  // not SERIAL_PORT_2
           case 27:  // M27
             printSetUpdateWaiting(false);
             break;
         #endif  // not SERIAL_PORT_2
 
-        case 73:
+        case 73:  // M73
```
Comment (kisslorand) [inline TFT/src/User/API/interfaceCmd.c:785]: Copy & Paste from PR https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/pull/2607

File: TFT/src/User/API/interfaceCmd.c [hunk 6]
```diff
@@ -1113,8 +1106,8 @@ void sendQueueCmd(void)
             setParameter(P_ABL_STATE, 1, cmd_float());
           break;
 
-        case 292:
-        case 408:
+        case 292:  // M292
```
Comment (kisslorand) [inline TFT/src/User/API/interfaceCmd.c:1109]: Copy & Paste from PR https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/pull/2607

File: TFT/src/User/API/interfaceCmd.c [hunk 7]
```diff
@@ -1113,8 +1106,8 @@ void sendQueueCmd(void)
...
+        case 408:  // M408
```
Comment (kisslorand) [inline TFT/src/User/API/interfaceCmd.c:1110]: Copy & Paste from PR https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/pull/2607

**Comment by digant73** [comment]
@kisslorand Please, have a break with such kind of complaints. Thanks

**Comment by kisslorand** [comment]
@bigtreetech
Please check #2685. Thank you!

**Comment by digant73** [comment]
Closing just to avoid any infinite loop. This PR will be eventually submitted/adapted in the future once kisslorand's constraints will be satisfied or rejected

**Comment by digant73** [comment]
Removed matching comments

**Comment by Evovita** [comment]
Thanks for this @digant73. 

After reading [Printing from host](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/#printing-from-host) notes in _README.md_ and adding `M118 P0 A1 action:pause` to Octoprint's `After print job is paused Printer > GCode` script I expected that the same Park/Pause gcode that runs if I press the Pause button on the TFT, would also run when _Octoprint_ Pause button is pressed. Instead, the print pauses, stuck on the model.

I guessing that I could add e.g. `M125` to the _Octoprint GCode_ to Park and pause, but from a logical perspective (generally speaking) the _Octoprint_ pause functionality should match the physical TFT pause functionality.

Do you know how this could be achieved without adding extra GCode to _Octoprint_?

**Pull request merged**